lib/kontena/actors/haproxy_spawner.rb

```diff
@@ -0,0 +1,56 @@
+
+module Kontena::Actors
+  class HaproxySpawner < Concurrent::Actor::RestartingContext
+    include Kontena::Logging
+
+    # @param [String] haproxy_bin
+    # @param [String] config_file
+    def initialize(haproxy_bin = '/usr/local/sbin/haproxy', config_file = '/etc/haproxy/haproxy.cfg')
+      @current_pid = nil
+      @haproxy_cmd = [haproxy_bin, '-f', config_file, '-db']
+      @validate_cmd = [haproxy_bin, '-c -f', config_file]
+    end
+
+    # @param [Kontena::Actors::Message] msg
+    def on_message(msg)
+      case msg.action
+      when :update
+        update_haproxy
+      else
+        pass
+      end
+    end
+
+    def update_haproxy
+      if validate_config
+        if current_pid
+          reload_haproxy
+        else
+          start_haproxy
+        end
+      end
+    end
+
+    def start_haproxy
+      info 'Starting HAProxy process'
+      @current_pid = Process.spawn(@haproxy_cmd.join(' '))
```

Comment:
Something must also be a `Process.wait` once haproxy exits? That happens after every reload, and this will leak haproxy zombies?

There should be a `SIGCHLD` handler somewhere?

Reply:
If haproxy process is reloaded with `-sf` it is allowed to process ongoing connections. So yes, it will leak processes until they drop all connections (and this is by design). Container also has `tini` running as process reaper.

We could probably improve this logic so that those old processes would be killed after a while (no point to keep them forever).